File: companies/Oracle/187.repeated-dna-sequences.java
import java.util.*;

/**
 * Algorithm:
 * Scan every length-10 substring. The first time a window repeats, add it to
 * the repeated set.
 *
 * Java data structures:
 * HashSet<String> stores seen windows and repeated windows.
 *
 * Complexity:
 * Time O(n) windows with constant length substring cost 10; space O(n).
 */
class Solution {
    public List<String> findRepeatedDnaSequences(String s) {
        Set<String> seen = new HashSet<>();
        Set<String> repeated = new HashSet<>();
        for (int i = 0; i + 10 <= s.length(); i++) {
            String window = s.substring(i, i + 10);
            if (!seen.add(window)) {
                repeated.add(window);
            }
        }
        return new ArrayList<>(repeated);
    }
}

File: Apple/187.repeated-dna-sequences.py
"""
Approach: Track seen 10-character windows and collect windows seen twice.
Data structure: two sets distinguish first sightings from already-reported repeated sequences.
Interview logic: every valid DNA substring has fixed length 10, so scanning all windows is enough. The output set prevents duplicates in the answer.
Complexity: O(n) time and O(n) space, with constant-size substring copies.
Tests and edge cases: strings shorter than 10 return []; overlapping repeats are allowed; a sequence repeated many times appears once.
"""
from __future__ import annotations
from typing import List

# @lc code=start
class Solution:
    def findRepeatedDnaSequences(self, s: str) -> List[str]:
        seen, repeated = set(), set()
        for i in range(len(s) - 9):
            window = s[i:i + 10]
            if window in seen:
                repeated.add(window)
            else:
                seen.add(window)
        return list(repeated)
# @lc code=end
